File: frontend/src/types/execution.ts
export type ExecutionStatus = "pending" | "running" | "completed" | "failed";

export interface Execution {
  id: number;
  template_id: number;
  template_name: string;
  provider: string;
  model: string;
  variables: Record<string, string>;
  status: ExecutionStatus;
  output?: string;
  error?: string;
  created_at: string;
  completed_at?: string;
}

export interface ExecutionCreateRequest {
  template_id: number;
  provider: string;
  model: string;
  variables: Record<string, string>;
}

export interface ExecutionListResponse {
  executions: Execution[];
  total: number;
}

export interface Provider {
  name: string;
  models: string[];
}

File: backend/app/models/execution.py
import enum
from datetime import datetime

from sqlalchemy import DateTime, ForeignKey, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import Base


class ExecutionStatus(str, enum.Enum):
    PENDING = "pending"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"


class Execution(Base):
    __tablename__ = "executions"

    id: Mapped[int] = mapped_column(primary_key=True)
    template_id: Mapped[int] = mapped_column(
        ForeignKey("templates.id"), nullable=False
    )
    template_version_id: Mapped[int] = mapped_column(
        ForeignKey("template_versions.id"), nullable=False
    )
    provider: Mapped[str] = mapped_column(String(50), nullable=False)
    model: Mapped[str] = mapped_column(String(100), nullable=False)
    variables: Mapped[str] = mapped_column(Text, nullable=False)
    status: Mapped[ExecutionStatus] = mapped_column(default=ExecutionStatus.PENDING)
    output: Mapped[str | None] = mapped_column(Text, nullable=True)
    error: Mapped[str | None] = mapped_column(Text, nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now()
    )
    completed_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )

    template: Mapped["Template"] = relationship()


from app.models.template import Template  # noqa: E402
